File: backend/src/schemas/gpx.py
"""
GPX schemas for GPS Routes feature.

Pydantic models for GPX file request/response validation in API.
Functional Requirements: FR-001 to FR-008, FR-036, FR-039
"""

from datetime import datetime

from pydantic import BaseModel, Field, field_validator

# ============================================================================
# Response Schemas (Output)
# ============================================================================


class TrackPointResponse(BaseModel):
    """
    Individual GPS trackpoint in the simplified route.

    Used in track visualization responses.
    """

    point_id: str = Field(..., description="Unique trackpoint identifier (UUID)")
    latitude: float = Field(..., ge=-90.0, le=90.0, description="Latitude in decimal degrees")
    longitude: float = Field(..., ge=-180.0, le=180.0, description="Longitude in decimal degrees")
    elevation: float | None = Field(None, description="Altitude in meters (NULL if not available)")
    distance_km: float = Field(..., ge=0.0, description="Cumulative distance from start in km")
    sequence: int = Field(..., ge=0, description="Order in track (0-indexed)")
    gradient: float | None = Field(None, description="Percentage slope (e.g., 5.2 = 5.2% uphill)")

    class Config:
        from_attributes = True


class CoordinateResponse(BaseModel):
    """
    Simple coordinate pair for route bounds.

    Used for start/end point markers.
    """

    latitude: float = Field(..., ge=-90.0, le=90.0, description="Latitude in decimal degrees")
    longitude: float = Field(..., ge=-180.0, le=180.0, description="Longitude in decimal degrees")


class GPXFileMetadata(BaseModel):
    """
    GPX file metadata and processing results.

    Used in GET /trips/{trip_id}/gpx response.
    """

    gpx_file_id: str = Field(..., description="Unique GPX file identifier (UUID)")
    trip_id: str = Field(..., description="Associated trip ID (UUID)")
    file_name: str = Field(..., description="Original filename (e.g., 'Camino_del_Cid.gpx')")
    file_size: int = Field(..., ge=0, description="File size in bytes")
    distance_km: float = Field(..., ge=0.0, description="Total distance in kilometers")
    elevation_gain: float | None = Field(None, description="Total elevation gain in meters")
    elevation_loss: float | None = Field(None, description="Total elevation loss in meters")
    max_elevation: float | None = Field(None, description="Maximum altitude in meters")
    min_elevation: float | None = Field(None, description="Minimum altitude in meters")
    total_points: int = Field(..., ge=0, description="Original trackpoint count")
    simplified_points: int = Field(
        ..., ge=0, description="Reduced trackpoint count after simplification"
    )
    has_elevation: bool = Field(..., description="True if GPX includes elevation data")
    has_timestamps: bool = Field(..., description="True if GPX includes timestamp data")
    processing_status: str = Field(
        ..., description="Current processing status (pending, processing, completed, error)"
    )
    error_message: str | None = Field(None, description="Error details if processing failed")
    uploaded_at: datetime = Field(..., description="Upload timestamp (ISO 8601)")
    processed_at: datetime | None = Field(None, description="Processing completion timestamp")

    class Config:
        from_attributes = True


class GPXUploadResponse(BaseModel):
    """
    Response from GPX upload endpoint.

    Two modes:
    - Synchronous (201): Small files (<1MB) processed immediately
    - Asynchronous (202): Large files (1-10MB) processing in background
    """

    gpx_file_id: str = Field(..., description="Unique GPX file identifier (UUID)")
    trip_id: str = Field(..., description="Associated trip ID (UUID)")
    processing_status: str = Field(
        ..., description="Current processing status (pending, processing, completed, error)"
    )

    # Fields below only present if processing_status = 'completed'
    distance_km: float | None = Field(None, description="Total distance (only if completed)")
    elevation_gain: float | None = Field(
        None, description="Total elevation gain (only if completed)"
    )
    elevation_loss: float | None = Field(
        None, description="Total elevation loss (only if completed)"
    )
    max_elevation: float | None = Field(None, description="Maximum altitude (only if completed)")
    min_elevation: float | None = Field(None, description="Minimum altitude (only if completed)")
    has_elevation: bool | None = Field(
        None, description="True if GPX has elevation (only if completed)"
    )
    has_timestamps: bool | None = Field(
        None, description="True if GPX has timestamps (only if completed)"
    )
    total_points: int | None = Field(
        None, description="Original trackpoint count (only if completed)"
    )
    simplified_points: int | None = Field(
        None, description="Reduced trackpoint count (only if completed)"
    )
    uploaded_at: datetime = Field(..., description="Upload timestamp (ISO 8601)")
    processed_at: datetime | None = Field(
        None, description="Processing completion timestamp (only if completed)"
    )
    error_message: str | None = Field(None, description="Error details (only if status=error)")

    class Config:
        from_attributes = True


class GPXStatusResponse(BaseModel):
    """
    Response from status polling endpoint.

    Used for async uploads to check processing progress.
    """

    gpx_file_id: str = Field(..., description="Unique GPX file identifier (UUID)")
    processing_status: str = Field(..., description="Current processing status")
    distance_km: float | None = Field(None, description="Total distance (only if completed)")
    elevation_gain: float | None = Field(
        None, description="Total elevation gain (only if completed)"
    )
    total_points: int | None = Field(
        None, description="Original trackpoint count (only if completed)"
    )
    simplified_points: int | None = Field(
        None, description="Reduced point count (only if completed)"
    )
    uploaded_at: datetime = Field(..., description="Upload timestamp")
    processed_at: datetime | None = Field(None, description="Processing completion timestamp")
    error_message: str | None = Field(None, description="Error details if processing failed")

    class Config:
        from_attributes = True


class TopClimbResponse(BaseModel):
    """
    Individual climb data in top climbs list.

    Represents a single climb segment with elevation and gradient details.
    Used in RouteStatisticsResponse.
    """

    start_km: float = Field(..., ge=0.0, description="Distance from start where climb begins (km)")
    end_km: float = Field(..., ge=0.0, description="Distance from start where climb ends (km)")
    elevation_gain_m: float = Field(..., ge=0.0, description="Total elevation gain in climb (meters)")
    avg_gradient: float = Field(
        ..., description="Average gradient of climb (percentage, e.g., 8.5 = 8.5% slope)"
    )
    description: str = Field(..., description="Human-readable climb description")

    @field_validator("end_km")
    @classmethod
    def end_km_must_be_greater_than_start(cls, v, info):
        """Validate that end_km > start_km."""
        if "start_km" in info.data and v <= info.data["start_km"]:
            raise ValueError("end_km must be greater than start_km")
        return v


class GradientCategoryResponse(BaseModel):
    """
    Statistics for one gradient category (FR-032).

    Used in gradient distribution visualization.
    """

    distance_km: float = Field(..., ge=0.0, description="Total distance in this category (km)")
    percentage: float = Field(..., ge=0.0, le=100.0, description="Percentage of total route distance")


class GradientDistributionResponse(BaseModel):
    """
    Complete gradient distribution breakdown (FR-032).

    Classifies route segments into gradient categories:
    - Llano (flat): 0-3%
    - Moderado (moderate): 3-6%
    - Empinado (steep): 6-10%
    - Muy empinado (very steep): >10%
    """

    llano: GradientCategoryResponse = Field(..., description="Flat terrain: 0-3% gradient")
    moderado: GradientCategoryResponse = Field(..., description="Moderate terrain: 3-6% gradient")
    empinado: GradientCategoryResponse = Field(..., description="Steep terrain: 6-10% gradient")
    muy_empinado: GradientCategoryResponse = Field(
        ..., description="Very steep terrain: >10% gradient"
    )


class RouteStatisticsResponse(BaseModel):
    """
    Advanced route statistics calculated from GPX data.

    Only present if GPX file has timestamps (has_timestamps=True).
    Includes speed metrics, time analysis, gradient analysis, and top climbs.
    """

    stats_id: str = Field(..., description="Unique statistics record ID (UUID)")
    gpx_file_id: str = Field(..., description="Associated GPX file ID (UUID)")
    avg_speed_kmh: float | None = Field(None, ge=0.0, lt=100.0, description="Average speed (km/h)")
    max_speed_kmh: float | None = Field(None, ge=0.0, lt=100.0, description="Maximum speed (km/h)")
    total_time_minutes: float | None = Field(None, ge=0.0, description="Total elapsed time (minutes)")
    moving_time_minutes: float | None = Field(
        None, ge=0.0, description="Time in motion, excludes stops (minutes)"
    )
    avg_gradient: float | None = Field(None, description="Average gradient over route (%)")
    max_gradient: float | None = Field(None, description="Maximum gradient (steepest uphill) (%)")
    top_climbs: list[TopClimbResponse] | None = Field(
        None, max_length=3, description="Top 3 hardest climbs (max 3 items)"
    )
    created_at: datetime = Field(..., description="When statistics were calculated (ISO 8601)")

    @field_validator("moving_time_minutes")
    @classmethod
    def moving_time_must_be_less_than_total(cls, v, info):
        """Validate that moving_time <= total_time."""
        if (
            v is not None
            and "total_time_minutes" in info.data
            and info.data["total_time_minutes"] is not None
        ):
            if v > info.data["total_time_minutes"]:
                raise ValueError("moving_time_minutes must be <= total_time_minutes")
        return v

    class Config:
        from_attributes = True


class RouteStatisticsWithDistributionResponse(RouteStatisticsResponse):
    """
    Extended route statistics with gradient distribution (FR-032).

    Includes all RouteStatisticsResponse fields plus detailed gradient breakdown.
    Used in frontend visualization for gradient distribution charts.
    """

    gradient_distribution: GradientDistributionResponse | None = Field(
        None, description="Gradient distribution breakdown (NULL if no elevation)"
    )


class TrackDataResponse(BaseModel):
    """
    Response from track data endpoint with simplified trackpoints.

    Used for map rendering and elevation profiles.
    Optionally includes advanced route statistics if timestamps are available.
    """

    gpx_file_id: str = Field(..., description="Unique GPX file identifier (UUID)")
    trip_id: str = Field(..., description="Associated trip ID (UUID)")
    distance_km: float = Field(..., ge=0.0, description="Total distance in kilometers")
    elevation_gain: float | None = Field(None, description="Total elevation gain in meters")
    simplified_points_count: int = Field(
        ..., ge=0, description="Number of points in trackpoints array"
    )
    has_elevation: bool = Field(..., description="True if trackpoints contain elevation data")
    start_point: CoordinateResponse = Field(..., description="Route start coordinate")
    end_point: CoordinateResponse = Field(..., description="Route end coordinate")
    trackpoints: list[TrackPointResponse] = Field(
        ..., description="Simplified trackpoints ordered by sequence"
    )
    route_statistics: RouteStatisticsResponse | None = Field(
        None, description="Advanced route statistics (only if GPX has timestamps)"
    )

    class Config:
        from_attributes = True


# ============================================================================
# Wrapper Schemas (Standard API Response Format)
# ============================================================================


class GPXUploadSuccessResponse(BaseModel):
    """
    Standard API response wrapper for GPX upload.

    Used in POST /trips/{trip_id}/gpx response.
    """

    success: bool = Field(True, description="Always true for successful uploads")
    data: GPXUploadResponse = Field(..., description="GPX upload data")
    message: str | None = Field(None, description="Optional message for async uploads")


class GPXMetadataSuccessResponse(BaseModel):
    """
    Standard API response wrapper for GPX metadata.

    Used in GET /trips/{trip_id}/gpx response.
    """

    success: bool = Field(True, description="Always true for successful requests")
    data: GPXFileMetadata = Field(..., description="GPX file metadata")


class GPXStatusSuccessResponse(BaseModel):
    """
    Standard API response wrapper for status polling.

    Used in GET /gpx/{gpx_file_id}/status response.
    """

    success: bool = Field(True, description="Always true for successful status checks")
    data: GPXStatusResponse = Field(..., description="GPX processing status")


class TrackDataSuccessResponse(BaseModel):
    """
    Standard API response wrapper for track data.

    Used in GET /gpx/{gpx_file_id}/track response.
    """

    success: bool = Field(True, description="Always true for successful requests")
    data: TrackDataResponse = Field(..., description="Track data with simplified points")

File: frontend/src/types/gpx.ts
/**
 * GPX TypeScript Type Definitions
 *
 * Type definitions for Feature 003 - GPS Routes Interactive.
 * Based on backend API contracts from specs/003-gps-routes/contracts/gpx-api.yaml.
 */

// ============================================================================
// Core Entities (API Responses)
// ============================================================================

/**
 * GPX file metadata and processing results
 *
 * Represents a GPX file attached to a trip, including:
 * - Original file metadata (url, size, name)
 * - Processing status (pending, processing, completed, error)
 * - Route statistics (distance, elevation, bounds)
 * - Track simplification metadata (total_points vs simplified_points)
 *
 * Used in:
 * - Trip detail view (display GPX stats)
 * - GPXUploader (upload status tracking)
 * - GPXMapView (render track on map)
 */
export interface GPXFileMetadata {
  /** Unique GPX file identifier (UUID) */
  gpx_file_id: string;

  /** Associated trip ID (UUID) */
  trip_id: string;

  /** Original filename (e.g., "Camino_del_Cid.gpx") */
  file_name: string;

  /** File size in bytes */
  file_size: number;

  /** Total distance in kilometers */
  distance_km: number;

  /** Total elevation gain in meters (null if no elevation data) */
  elevation_gain: number | null;

  /** Total elevation loss in meters (null if no elevation data) */
  elevation_loss: number | null;

  /** Maximum altitude in meters (null if no elevation data) */
  max_elevation: number | null;

  /** Minimum altitude in meters (null if no elevation data) */
  min_elevation: number | null;

  /** Original trackpoint count before simplification */
  total_points: number;

  /** Reduced trackpoint count after Douglas-Peucker simplification */
  simplified_points: number;

  /** True if GPX includes elevation data */
  has_elevation: boolean;

  /** True if GPX includes timestamp data */
  has_timestamps: boolean;

  /** Current processing status */
  processing_status: GPXProcessingStatus;

  /** Error details if processing failed */
  error_message: string | null;

  /** Upload timestamp (ISO 8601) */
  uploaded_at: string;

  /** Processing completion timestamp (ISO 8601, null if not yet processed) */
  processed_at: string | null;
}

/**
 * Individual GPS trackpoint with coordinates and metrics
 *
 * Represents a single point on the simplified GPS track with:
 * - Geographic coordinates (latitude, longitude, elevation)
 * - Cumulative distance from route start
 * - Sequence number for ordering
 * - Calculated gradient (% slope)
 *
 * Used in:
 * - GPXMapView (render polyline on map)
 * - ElevationProfile (render elevation chart)
 * - Track analytics
 */
export interface TrackPoint {
  /** Unique trackpoint identifier (UUID) */
  point_id: string;

  /** Latitude in decimal degrees (-90 to 90) */
  latitude: number;

  /** Longitude in decimal degrees (-180 to 180) */
  longitude: number;

  /** Altitude in meters (null if not available) */
  elevation: number | null;

  /** Cumulative distance from start in kilometers */
  distance_km: number;

  /** Order in track (0-indexed) */
  sequence: number;

  /** Percentage slope (e.g., 5.2 = 5.2% uphill, null if no elevation data) */
  gradient: number | null;
}

/**
 * Geographic coordinate (start/end point)
 *
 * Simple lat/lon coordinate pair for route bounds.
 *
 * Used in:
 * - GPXMapView (display start/end markers)
 * - Track bounds calculation
 */
export interface Coordinate {
  /** Latitude in decimal degrees (-90 to 90) */
  latitude: number;

  /** Longitude in decimal degrees (-180 to 180) */
  longitude: number;
}

// ============================================================================
// Response Schemas
// ============================================================================

/**
 * Response from GPX upload endpoint
 *
 * Returned by POST /trips/{trip_id}/gpx after uploading a GPX file.
 *
 * Two modes:
 * - Synchronous (201): Small files (<1MB) processed immediately, includes full metadata
 * - Asynchronous (202): Large files (1-10MB) processing in background, includes status_url
 *
 * Used in:
 * - GPXUploader (upload feedback)
 * - useGPXUpload hook
 */
export interface GPXUploadResponse {
  /** Always true for successful uploads */
  success: true;

  /** GPX file data */
  data: {
    /** Unique GPX file identifier (UUID) */
    gpx_file_id: string;

    /** Associated trip ID (UUID) */
    trip_id: string;

    /** Current processing status */
    processing_status: GPXProcessingStatus;

    // Fields below only present if processing_status = 'completed'

    /** Total distance in kilometers (only if completed) */
    distance_km?: number;

    /** Total elevation gain in meters (only if completed) */
    elevation_gain?: number | null;

    /** Total elevation loss in meters (only if completed) */
    elevation_loss?: number | null;

    /** Maximum altitude in meters (only if completed) */
    max_elevation?: number | null;

    /** Minimum altitude in meters (only if completed) */
    min_elevation?: number | null;

    /** True if GPX includes elevation data (only if completed) */
    has_elevation?: boolean;

    /** True if GPX includes timestamp data (only if completed) */
    has_timestamps?: boolean;

    /** Original trackpoint count (only if completed) */
    total_points?: number;

    /** Reduced trackpoint count (only if completed) */
    simplified_points?: number;

    /** Upload timestamp (ISO 8601) */
    uploaded_at: string;

    /** Processing completion timestamp (ISO 8601, only if completed) */
    processed_at?: string;

    /** Error details (only if processing_status = 'error') */
    error_message?: string;
  };

  /** Optional message for async uploads */
  message?: string;
}

/**
 * Response from GPX status polling endpoint
 *
 * Returned by GET /gpx/{gpx_file_id}/status when polling async uploads.
 *
 * Poll every 2 seconds until status = "completed" or "error".
 *
 * Used in:
 * - GPXUploader (async upload polling)
 * - useGPXUpload hook
 */
export interface GPXStatusResponse {
  /** Always true for successful status checks */
  success: true;

  /** Same structure as GPXUploadResponse.data */
  data: GPXUploadResponse['data'];
}

/**
 * Individual climb data in top climbs list (User Story 5)
 *
 * Represents a single climb segment with elevation and gradient details.
 */
export interface TopClimb {
  /** Distance from start where climb begins (km) */
  start_km: number;

  /** Distance from start where climb ends (km) */
  end_km: number;

  /** Length of climb (km) - calculated as end_km - start_km */
  distance_km: number;

  /** Total elevation gain in climb (meters) */
  elevation_gain_m: number;

  /** Average gradient of climb (percentage, e.g., 8.5 = 8.5% slope) */
  avg_gradient: number;

  /** Human-readable climb description */
  description: string;
}

/**
 * Gradient category statistics (User Story 5 - FR-032)
 *
 * Statistics for one gradient category in the route.
 */
export interface GradientCategory {
  /** Total distance in this category (km) */
  distance_km: number;

  /** Percentage of total route distance */
  percentage: number;
}

/**
 * Gradient distribution breakdown (User Story 5 - FR-032)
 *
 * Classifies route segments into gradient categories:
 * - Llano (flat): 0-3%
 * - Moderado (moderate): 3-6%
 * - Empinado (steep): 6-10%
 * - Muy empinado (very steep): >10%
 */
export interface GradientDistribution {
  /** Flat terrain: 0-3% gradient */
  llano: GradientCategory;

  /** Moderate terrain: 3-6% gradient */
  moderado: GradientCategory;

  /** Steep terrain: 6-10% gradient */
  empinado: GradientCategory;

  /** Very steep terrain: >10% gradient */
  muy_empinado: GradientCategory;
}

/**
 * Advanced route statistics (User Story 5)
 *
 * Only present if GPX file has timestamps (has_timestamps=True).
 * Includes speed metrics, time analysis, gradient analysis, and top climbs.
 */
export interface RouteStatistics {
  /** Unique statistics record ID (UUID) */
  stats_id: string;

  /** Associated GPX file ID (UUID) */
  gpx_file_id: string;

  /** Average speed (km/h) */
  avg_speed_kmh: number | null;

  /** Maximum speed (km/h) */
  max_speed_kmh: number | null;

  /** Total elapsed time (minutes) */
  total_time_minutes: number | null;

  /** Time in motion, excludes stops (minutes) */
  moving_time_minutes: number | null;

  /** Average gradient over route (%) */
  avg_gradient: number | null;

  /** Maximum gradient (steepest uphill) (%) */
  max_gradient: number | null;

  /** Top 3 hardest climbs (max 3 items) */
  top_climbs: TopClimb[] | null;

  /** When statistics were calculated (ISO 8601) */
  created_at: string;

  /** Gradient distribution breakdown (FR-032) */
  gradient_distribution?: GradientDistribution;
}

/**
 * Response from track data endpoint
 *
 * Returned by GET /gpx/{gpx_file_id}/track with simplified trackpoints for rendering.
 *
 * Used in:
 * - GPXMapView (render route polyline)
 * - ElevationProfile (render elevation chart)
 * - useGPXTrack hook
 */
export interface TrackDataResponse {
  /** Always true for successful requests */
  success: true;

  /** Track data with simplified points */
  data: {
    /** Unique GPX file identifier (UUID) */
    gpx_file_id: string;

    /** Associated trip ID (UUID) */
    trip_id: string;

    /** Total distance in kilometers */
    distance_km: number;

    /** Total elevation gain in meters (null if no elevation data) */
    elevation_gain: number | null;

    /** Number of points in trackpoints array */
    simplified_points_count: number;

    /** True if trackpoints contain elevation data */
    has_elevation: boolean;

    /** Route start coordinate */
    start_point: Coordinate;

    /** Route end coordinate */
    end_point: Coordinate;

    /** Simplified trackpoints ordered by sequence */
    trackpoints: TrackPoint[];

    /** Advanced route statistics (only if GPX has timestamps) - User Story 5 */
    route_statistics: RouteStatistics | null;
  };
}

// ============================================================================
// Enums and Constants
// ============================================================================

/**
 * GPX file processing status
 *
 * Lifecycle:
 * - pending: File uploaded, not yet processing
 * - processing: Background task running (for files >1MB)
 * - completed: Successfully processed, trackpoints extracted
 * - error: Processing failed (malformed GPX, corrupted file, etc.)
 */
export type GPXProcessingStatus = 'pending' | 'processing' | 'completed' | 'error';

/**
 * Processing status labels in Spanish
 */
export const PROCESSING_STATUS_LABELS: Record<GPXProcessingStatus, string> = {
  pending: 'Pendiente',
  processing: 'Procesando',
  completed: 'Completado',
  error: 'Error',
};

/**
 * Processing status CSS class names
 */
export const PROCESSING_STATUS_CLASSES: Record<GPXProcessingStatus, string> = {
  pending: 'gpx-status--pending',
  processing: 'gpx-status--processing',
  completed: 'gpx-status--completed',
  error: 'gpx-status--error',
};

// ============================================================================
// Component Props (Client-Only)
// ============================================================================

/**
 * Props for GPXUploader component
 *
 * Used in:
 * - TripDetailPage (upload GPX to trip)
 */
export interface GPXUploaderProps {
  /** Trip ID to attach GPX file to */
  tripId: string;

  /** Callback when upload completes successfully */
  onUploadComplete?: (gpxFileId: string) => void;

  /** Callback when upload fails */
  onUploadError?: (error: string) => void;
}

/**
 * Props for GPXMapView component
 *
 * Used in:
 * - TripDetailPage (display route on map)
 */
export interface GPXMapViewProps {
  /** Trackpoints to render as polyline */
  trackPoints: TrackPoint[];

  /** Route start coordinate */
  startPoint: Coordinate;

  /** Route end coordinate */
  endPoint: Coordinate;

  /** Auto-fit map bounds to track on load (default: true) */
  autoFitBounds?: boolean;

  /** Callback when trackpoint is clicked */
  onPointClick?: (point: TrackPoint) => void;
}

/**
 * Props for ElevationProfile component
 *
 * Used in:
 * - TripDetailPage (display elevation chart)
 */
export interface ElevationProfileProps {
  /** Trackpoints to render in elevation profile */
  trackPoints: TrackPoint[];

  /** Callback when chart point is clicked */
  onPointClick?: (point: TrackPoint) => void;

  /** Chart height in pixels (default: 300) */
  height?: number;
}

/**
 * Props for GPXStats component
 *
 * Used in:
 * - TripDetailPage (display route statistics)
 */
export interface GPXStatsProps {
  /** GPX file metadata to display */
  metadata: GPXFileMetadata;

  /** GPX file ID (required for download functionality) */
  gpxFileId?: string;

  /** Whether current user is trip owner (enables download and delete buttons) */
  isOwner?: boolean;

  /** Callback when GPX file is deleted (owner-only action) */
  onDelete?: () => void;
}

// ============================================================================
// Upload State (Client-Only)
// ============================================================================

/**
 * GPX upload state for useGPXUpload hook
 *
 * Tracks upload progress and processing status.
 *
 * Used in:
 * - GPXUploader component
 * - useGPXUpload hook
 */
export interface GPXUploadState {
  /** Whether upload is in progress */
  isUploading: boolean;

  /** Whether processing is in progress (for async uploads) */
  isProcessing: boolean;

  /** Upload progress percentage (0-100) */
  uploadProgress: number;

  /** Current processing status */
  processingStatus: GPXProcessingStatus | null;

  /** Uploaded GPX file ID (null until upload completes) */
  gpxFileId: string | null;

  /** Error message (null if no error) */
  error: string | null;
}
